### PR TITLE
Celeste 64: v1.2 Content Update

### PR DESCRIPTION
**Pull request by PoryGone**
Changelog:

```md
## Features:

- New optional Location Checks
	- Friendsanity
	- Signsanity
	- Carsanity
- Move Shuffle
	- Basic movement abilities can be shuffled into the item pool
		- Ground Dash
		- Air Dash
		- Skid Jump
		- Climb
- Logic Difficulty
	- Completely overhauled logic system
	- Standard or Hard logic difficulty can be chosen
- Badeline Chasers
	- Opt-in options which cause Badelines to start following you as you play, which will kill on contact
	- These can be set to spawn based on either:
		- The number of locations you've checked
		- The number of Strawberry items you've received
	- How fast they follow behind you can be specified

## Quality of Life:

- The maximum number of Strawberries in the item pool can be directly set
	- The required amount of Strawberries is now set via percentage
	- All items beyond the amount placed in the item pool will be `Raspberry` items, which have no effect
- Any unique items placed into the `start_inventory` will not be placed into the item pool
```

Accompanying game release can be found [here](https://github.com/PoryGoneDev/Celeste64/releases/tag/v1.2.0).

### CODE REVIEW

File: worlds/celeste64/Options.py
Comment (Exempt-Medic) [inline worlds/celeste64/Options.py:9]: Perhaps this is out of scope for this change, but I think it's worth pointing out that the other two worlds with DeathLink Amnesty both have it as a range starting at 0. Unsure if you'd want to change it, but I figured you should know

File: worlds/celeste64/__init__.py
Comment (Exempt-Medic) [inline worlds/celeste64/__init__.py:78]: I think explaining this behavior somewhere would be a good idea. Like in the option description for `logic_difficulty`

File: worlds/celeste64/Rules.py
Comment (Exempt-Medic) [inline worlds/celeste64/Rules.py:18]:
```suggestion
goal_standard_logic: List[List[str]] = [[ItemName.feather, ItemName.traffic_block, ItemName.breakables, ItemName.double_dash_refill]]
```

**Review by Exempt-Medic — CHANGES_REQUESTED** [review]
Looked through the code, merged into `main` and did a few thousand generations without any issues. I don't know enough about the game to review the logic of `rules.py`. Made a few suggested changes

File: worlds/celeste64/Options.py
Comment (Exempt-Medic) [reply worlds/celeste64/Options.py:9]: "Amnesty" usually means being forgiven/pardoned. So setting your "forgiveness" to 1 and that being the _minimum_ doesn't really make sense. Looking at how the other games actually word the option though, I would say they don't make sense either. I think "Amnesty" may just not be a good word for what's trying to be conveyed here

Comment (Exempt-Medic) [reply worlds/celeste64/Options.py:9]: And given that the word is "amnesty" one would assume the minimum amount of forgiveness you can have is none, which maps much better to 0 than to 1

Comment (Exempt-Medic) [reply worlds/celeste64/Options.py:9]: You can mark this as resolved then

File: worlds/celeste64/Options.py
Comment (Exempt-Medic) [reply worlds/celeste64/Options.py:82]: What yamls would be breaking? Isn't this a new option in this version?

Comment (Exempt-Medic) [reply worlds/celeste64/Options.py:97]: You can use an option_alias then

Comment (Exempt-Medic) [reply worlds/celeste64/Options.py:97]: You could use a hidden/removed option to throw an error if the old option is used

File: worlds/celeste64/__init__.py
Comment (Exempt-Medic) [reply worlds/celeste64/__init__.py:78]: LGTM

**Review by Exempt-Medic — APPROVED** [review]
Merged into `main`, lots of test generations. Didn't look at `rules.py` much and I'm wondering if `location_rule` could be reformated to have less checking of the logic difficulty / move shuffling, but otherwise it looks good

File: worlds/celeste64/Rules.py
Comment (NewSoupVi) [inline worlds/celeste64/Rules.py:323]: For performance, I would very much prefer if you moved as many conditions out of the actual access rule evaluation as possible, and did them ahead of time.
This if/else is going to be called each time an item is placed, so 1000s or 10000s of times, when it could be done once 

The simplest change I can think of is moving `active_location_mapping` to an attribute of e.g. `Celeste64World` that you set once.
So, move this whole if/else block somewhere else and save the attribute to the world (PyCharm might complain about "setting attribute outside of `__init__`, maybe you can figure something out)

```
def set_rules(world: Celeste64World):
    if world.options.logic_difficulty == "standard":
        if world.options.move_shuffle:
            world.active_logic_mapping = location_standard_moves_logic
        else:
            world.active_logic_mapping = location_standard_logic
    else:
        if world.options.move_shuffle:
            world.active_logic_mapping = location_hard_moves_logic
        else:
            world.active_logic_mapping = location_hard_logic

    ...
```

and then in location_rule, you can just have:

```
def location_rule(state: CollectionState, world: Celeste64World, loc: str) -> bool:
    if loc not in world.active_logic_mapping:
        return True

    for possible_access in world.active_logic_mapping[loc]:
        if state.has_all(possible_access, world.player):
            return True

    return False
```

At this point, I'd be willing to approve the PR (after reviewing everything else as well)
But, there are more optimisations to be made that I would recommend that I will detail here as well.

Now that we have the active location mapping at an earlier point, instead of this "in" check inside the location_rule function, you could just do

```
def set_rules(world: Celeste64World):
    for location in world.multiworld.get_locations(world.player):
        if location in world.active_location_mapping:
            set_rule(location, lambda state, location=location: location_rule(state, world, location.name))
        else:
            set_rule(location, lambda state: True)
```

With location_rule now being:

```
def location_rule(state: CollectionState, world: Celeste64World, loc: str) -> bool:
    for possible_access in world.active_logic_mapping[loc]:
        if state.has_all(possible_access, world.player):
            return True

    return False
```

And finally, we can replace this early exiting for loop with an `any` (which also inherently exits early)

```
def location_rule(state: CollectionState, world: Celeste64World, loc: str) -> bool:
    return any(
        state.has_all(possible_access, world.player)
        for possible_access in world.active_logic_mapping[loc]
   )
```

**Review by NewSoupVi — APPROVED** [review]
Some earlier evalutions of the conditions inside the location_rule function would be quite important to me personally.

**Review by NewSoupVi — CHANGES_REQUESTED** [review]
Clicked the wrong damn button again

File: worlds/celeste64/Rules.py
Comment (NewSoupVi) [reply worlds/celeste64/Rules.py:323]: I'll give my thoughts on that, sure. Let me put the "size of the game" thing aside for a second, I'll come back to it.

So, I've implemented the changes I suggested and I'm observing about a 30% speedup. It's not crazy good or anything, but considering:
1. These suggestions are fairly simple (in the sense that they don't increase overall line count and still keep all of the same code, just shuffled around)
2. this is a *change* you're making from the original implementation (i.e. it is "in scope" of the PR)
3. "Avoid option checking inside of access rules" **is** a common piece of advice given by multiple world and core maintainers, like I'm not just pulling this out of my backside, there is some "consensus" around this - I disagree with the wording choice of "minutia" here because of this

It is something I will feel inclined to comment on.

Now, the size of the game might be a good argument, and I think I'm "emotionally" willing to let it go? But **please please please** consider the first suggestion I made, it is very simple and already gets 25% of a speedup, everything else is in the 1%-2% range.

I will however say that I'm not sure I'm **generically** comfortable with the idea - and you didn't say this, but this would be my takeaway from this conversation - that I should take the size of a game into consideration when reviewing PRs like this (new game or big systems refactor). This problem is something I'd absolutely be calling out and not letting fly for a 500 location game - Is it that fair or reasonable for me to be softer on it when the game is smaller? I'm not sure about that

File: just_gen500.bat
Comment (NewSoupVi) [inline just_gen500.bat:3]: Think this got in here by accident lol

**Review by NewSoupVi — CHANGES_REQUESTED** [review]

**Review by NewSoupVi — APPROVED** [review]
Code LGTM, very clean

Before I hit the button, I think I would like either a review from someone who knows the game, or a record that playtests were run / the new version has successfully been played in unsupported games.
Mostly to make sure that the logic stuff is correct, as neither Exempt-Medic (I think) nor I know anything about the game and have purely looked at this from a code quality / generation perspective

If playtests have already been run, then we're good obviously, just let me know :)

**Comment by NewSoupVi** [comment]
Good enough for me :)

**Pull request merged**
